frame/support/src/dispatch.rs

```diff
@@ -542,6 +552,98 @@ macro_rules! decl_module {
 	) => {
 		compile_error!("`on_finalize` can only be passed once as input.");
 	};
+
+	// Add on_idle
+	(@normalize
+		$(#[$attr:meta])*
+		pub struct $mod_type:ident<$trait_instance:ident: $trait_name:ident$(<I>, I: $instantiable:path $(= $module_default_instance:path)?)?>
+		for enum $call_type:ident where origin: $origin_type:ty, system = $system:ident
+		{ $( $other_where_bounds:tt )* }
+		{ $( $deposit_event:tt )* }
+		{ $( $on_initialize:tt )* }
+		{ $( $on_runtime_upgrade:tt )* }
+		{}
+		{ $( $on_finalize:tt )* }
+		{ $( $offchain:tt )* }
+		{ $( $constants:tt )* }
+		{ $( $error_type:tt )* }
+		{ $( $integrity_test:tt )* }
+		[ $( $dispatchables:tt )* ]
+		$(#[doc = $doc_attr:tt])*
+		fn on_idle( $( $param_name:ident : $param:ty ),* $(,)? ) { $( $impl:tt )* }
```

Comment:
We should parse the error. Also I think we can force user to give both argument.
```suggestion
		fn on_idle($param_name1:ident : $param1:ty, $param_name2:ident: $param2:ty $(,)? ) -> $return:ty { $( $impl:tt )* }
```
But honestly I was also ok not allowing decl_module to define its own on_idle.

Reply:
I actually have this exact change stashed, I even call it param 2 too lol. However it does seem to be a bit challenging. So I need to have on idle implemented here, but if we want to not make it available here I guess just one compile error that says something like on_idle is only available in frame V3?

Reply:
yes to make a compile error when on_idle is used we can do:
```rust
	(@normalize
		$(#[$attr:meta])*
		pub struct $mod_type:ident<
			$trait_instance:ident: $trait_name:ident$(<I>, I: $instantiable:path $(= $module_default_instance:path)?)?
		>
		for enum $call_type:ident where origin: $origin_type:ty, system = $system:ident
		{ $( $other_where_bounds:tt )* }
		{ $( $deposit_event:tt )* }
		{ $( $on_initialize:tt )* }
		{ $( $on_runtime_upgrade:tt )* }
		{ $( $on_finalize:tt )* }
		{ $( $offchain:tt )* }
		{ $( $constants:tt )* }
		{ $( $error_type:tt )* }
		{ $( $integrity_test:tt )* }
		[ $( $dispatchables:tt )* ]
		$(#[doc = $doc_attr:tt])*
		$(#[weight = $weight:expr])?
		fn on_idle
		$($rest:tt)*
	) => {
		compile_error!("`on_idle` cannot be implemented in decl_module, name is reserved to avoid confusion with on_idle hook");
	};
```